pallets/liquidity-pools/src/hooks.rs

```diff
@@ -0,0 +1,109 @@
+use sp_std::marker::PhantomData;
+
+use crate::{pallet::Config, Message, MessageOf, Pallet};
+
+/// The hook struct which acts upon a finalized investment decrement.
+pub struct DecreaseInvestOrderHook<T>(PhantomData<T>);
+
+/// The hook struct which acts upon a finalized redemption collection.
+
+pub struct CollectRedeemHook<T>(PhantomData<T>);
+
+impl<T: Config> StatusNotificationHook for DecreaseInvestOrderHook<T>
+where
+	<T as frame_system::Config>::AccountId: Into<[u8; 32]>,
+{
+	type Error = DispatchError;
+	type Id = ForeignInvestmentInfo<T::AccountId, T::TrancheCurrency, ()>;
+	type Status = ExecutedForeignDecrease<T::Balance, T::CurrencyId>;
+
+	#[transactional]
+	fn notify_status_change(
+		id: ForeignInvestmentInfo<T::AccountId, T::TrancheCurrency, ()>,
+		status: ExecutedForeignDecrease<T::Balance, T::CurrencyId>,
+	) -> DispatchResult {
+		let ForeignInvestmentInfo {
+			id: investment_id,
+			owner: investor,
+			..
+		} = id;
+		let currency = Pallet::<T>::try_get_general_index(status.return_currency)?;
+		let wrapped_token = Pallet::<T>::try_get_wrapped_token(&status.return_currency)?;
+		let domain_address: DomainAddress = wrapped_token.into();
+
+		T::Tokens::burn_from(status.return_currency, &investor, status.amount_decreased)?;
+
+		let message: MessageOf<T> = Message::ExecutedDecreaseInvestOrder {
+			pool_id: investment_id.of_pool(),
+			tranche_id: investment_id.of_tranche(),
+			investor: investor.clone().into(),
+			currency,
+			currency_payout: status.amount_decreased,
+		};
+
+		T::OutboundQueue::submit(investor, domain_address.domain(), message)?;
+
+		Ok(())
+	}
+}
+
+impl<T: Config> StatusNotificationHook for CollectRedeemHook<T>
+where
+	<T as frame_system::Config>::AccountId: Into<[u8; 32]>,
+{
+	type Error = DispatchError;
+	type Id = ForeignInvestmentInfo<T::AccountId, T::TrancheCurrency, ()>;
+	type Status = cfg_types::investments::ExecutedForeignCollectRedeem<T::Balance, T::CurrencyId>;
+
+	#[transactional]
+	fn notify_status_change(
+		id: ForeignInvestmentInfo<T::AccountId, T::TrancheCurrency, ()>,
+		status: cfg_types::investments::ExecutedForeignCollectRedeem<T::Balance, T::CurrencyId>,
+	) -> DispatchResult {
+		let ForeignInvestmentInfo {
+			id: investment_id,
+			owner: investor,
+			..
+		} = id;
+		let currency = Pallet::<T>::try_get_general_index(status.currency)?;
+		let wrapped_token = Pallet::<T>::try_get_wrapped_token(&status.currency)?;
+		let domain_address: DomainAddress = wrapped_token.into();
+
+		T::Tokens::burn_from(status.currency, &investor, status.amount_currency_payout)?;
+
+		let message: MessageOf<T> = Message::ExecutedCollectInvest {
+			pool_id: investment_id.of_pool(),
+			tranche_id: investment_id.of_tranche(),
+			investor: investor.clone().into(),
+			currency,
+			currency_payout: status.amount_currency_payout,
+			tranche_tokens_payout: status.amount_tranche_tokens_payout,
+		};
+
+		T::OutboundQueue::submit(investor, domain_address.domain(), message)?;
```

Comment:
investor will need to pay if we use them as the sender. Not sure, if we should maybe use the treasury account here. In most case I would assume, that investors from other domain do not hold CFG in their derived address.

(same as above)